File: CLearn/CLearn/DataType.c
//
//  DataType.c
//  CLearn
//
//  Created by 翁恒丛 on 2018/11/8.
//  Copyright © 2018年 翁恒丛. All rights reserved.
//

#include "DataType.h"

void logDataTypeMemorySize()
{
    printf("在64位机器中：\n");
    printf("        char类型----%lu个字节\n",sizeof(char));
    printf("       short类型----%lu个字节\n",sizeof(short));
    printf("         int类型----%lu个字节\n",sizeof(int));
    printf("        long类型----%lu个字节\n",sizeof(long));
    printf("unsigned int类型----%lu个字节\n",sizeof(unsigned int));
    printf("       float类型----%lu个字节\n",sizeof(float));
    printf("      double类型----%lu个字节\n",sizeof(double));
    printf(" long double类型----%lu个字节\n",sizeof(long double));
    printf("-------------------指针----4个字节---------------------\n");
    char *p0;
    short *p1;
    int *p2;
    long *p3;
    unsigned int *p4;
    float *p5;
    double *p6;
    long double *p7;
    printf("        char类型指针----%lu个字节\n",sizeof(p0));
    printf("       short类型指针----%lu个字节\n",sizeof(p1));
    printf("         int类型指针----%lu个字节\n",sizeof(p2));
    printf("        long类型指针----%lu个字节\n",sizeof(p3));
    printf("unsigned int类型指针----%lu个字节\n",sizeof(p4));
    printf("       float类型指针----%lu个字节\n",sizeof(p5));
    printf("      double类型指针----%lu个字节\n",sizeof(p6));
    printf(" long double类型指针----%lu个字节\n",sizeof(p7));
}

File: 算法/算法题练习/算法题练习/Base/DataType.m
//
//  DataType.m
//  算法题练习
//
//  Created by WengHengcong on 2019/1/25.
//  Copyright © 2019 WengHengcong. All rights reserved.
//

#import "DataType.h"

@implementation ListNode

- (instancetype)initWithValue:(id)value
{
    if (self = [self init]) {
        self.value = value;
        self.next = nil;
        self.prev = nil;
    }
    return self;
}

+ (instancetype)nodeWithValue:(id)value
{
    return [[self alloc] initWithValue:value];
}

- (id)copyWithZone:(NSZone *)zone
{
    // 将属性全部声明为copy
    ListNode *node = [[ListNode alloc] init];
    node.next = self.next;  //保持链接关系
    node.prev = self.prev;  //保持链接关系
    node.value = [self.value copy];
    return node;
}

@end

@implementation TreeNode

- (instancetype)initWithKey:(nonnull id)key value:(id)value size:(int)size
{
    self = [super init];
    self.key = key;
    self.value = value;
    self.size = size;
    return self;
}

+ (instancetype)nodeWithKey:(nonnull id)key value:(id)value size:(int)size
{
    return [[self alloc] initWithKey:key value:value size:size];
}

- (id)copyWithZone:(NSZone *)zone
{
    // 将属性全部声明为copy
    TreeNode *node = [[TreeNode alloc] init];
    node.left = self.left;  //保持链接关系
    node.right = self.right;  //保持链接关系
    node.key = self.key;
    node.value = [self.value copy];
    node.size = self.size;
    return node;
}

@end
